R/utils.R

```diff
@@ -132,8 +133,12 @@ eval_with_cols = function(orig_call, all_cols) {
   })
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
-      named_call[["cols"]] = all_cols
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
+        stopf("user should not provide cols argument to %s, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument", as.character(fun_uneval))
+      } else {
+        named_call[["cols"]] = all_cols
+      }
```

Comment:
I call the `else` branch here "unnecessary nesting" -- the `stop()` branch will always exit, so we can reduce the indentation of the non`-stop()` branch. (NB: this is something caught by `lintr::unnecessary_nesting_linter()`, we can address this generally in the repo if we so choose.)

```suggestion
      }
      named_call[["cols"]] = all_cols
```